### PR TITLE
TATB : 06 Change the JSON file with Proper Format and get Options randomly in the quiz.

### PR DESCRIPTION
**Pull request by PrabhatJTDeveloper**

### CODE REVIEW

File: src/app/component/quiz/quiz.component.html [hunk 1]
```diff
@@ -1,53 +1,37 @@
...
                       </div>
-                      <div class="ram">
-                        <i
-                          style="font-size: 20px; width: 80px"
-                          class="fa fa-clock"
+                      <div class="timer">
+                        <i class="fa fa-clock timer"
                           >&#xf017;
                           {{ FormArray.controls.at(i)?.value?.timer }} sec</i
                         >
```
Comment (PoojanPatelJT) [inline src/app/component/quiz/quiz.component.html]: Change FormArray to formArray

File: src/app/component/quiz/quiz.component.html [hunk 2]
```diff
@@ -1,53 +1,37 @@
...
                       </div>
                     </div>
 
                     <ng-container
-                      *ngFor="let card of question[this.questionindex].options"
+                      *ngFor="let card of question[this.questionIndex].options"
                     >
```
Comment (PoojanPatelJT) [inline src/app/component/quiz/quiz.component.html]: Do we use i as index for looping

File: src/app/component/quiz/quiz.component.ts [hunk 3]
```diff
@@ -98,11 +70,27 @@ export class Quizcomponent implements OnInit, OnDestroy {
...
+    )?.numberOfQuestions;
+    let data: any = this.quizData.quiz.find(
+      (data) => data.quizId == this.selectedQuizType
+    )?.questions;
+    this.question = [...data];
     this.question = this.question?.sort(() => Math.random() - 0.67);
-    this.question = [...this.question?.splice(0, 11)];
-    this.options = this.question[this.questionindex]?.options;
-    this.timer = this.quizData[this.selectedQuizType]?.timer;
+    this.question = [...this.question?.splice(0, this.numberOfQuestions)];
```
Comment (PoojanPatelJT) [inline src/app/component/quiz/quiz.component.ts]: Please create a object and then use it directly

File: src/app/component/quiz/quiz.component.ts [hunk 4]
```diff
@@ -119,23 +107,23 @@ export class Quizcomponent implements OnInit, OnDestroy {
     return this.myForm.controls['form'] as FormArray;
   }
 
-  next(questionindex: number): void {
+  nextQuestion(questionIndex: number): void {
```
Comment (PoojanPatelJT) [inline src/app/component/quiz/quiz.component.ts]: why using void ?

File: src/app/component/quiz/quiz.component.ts [hunk 5]
```diff
@@ -156,75 +144,67 @@ export class Quizcomponent implements OnInit, OnDestroy {
     this.router.navigate(['/result']);
```
Comment (PoojanPatelJT) [inline src/app/component/quiz/quiz.component.ts]: please use navigateByUrl

File: src/app/component/quiz/quiz.component.ts [hunk 6]
```diff
@@ -156,75 +144,67 @@ export class Quizcomponent implements OnInit, OnDestroy {
...
-      }
-      let selectedAnswer =
-        this.question[questionindex].answer?.[0] == correctOptions;
-      if (selectedAnswer && correctOptions) {
-        this.points = this.points += 1;
+  answer(questionIndex: number, correctOptions: string) {
+    if (!this.FormArray.at(questionIndex).get('timer')?.disabled) {
+      if (this.question[questionIndex].answer.id && correctOptions) {
+        this.points = this.points += this.positivePoints;
         this.correctanswer++;
```
Comment (PoojanPatelJT) [inline src/app/component/quiz/quiz.component.ts]: Use camelcase

File: src/app/component/quiz/quiz.component.ts [hunk 7]
```diff
@@ -156,75 +144,67 @@ export class Quizcomponent implements OnInit, OnDestroy {
...
-      } else if (!selectedAnswer && correctOptions) {
-        debugger;
-        this.points = this.points -= 0.25;
+      } else if (!this.question[questionIndex].answer.id && correctOptions) {
+        this.points = this.points -= this.negativePoints;
         this.inCorrectanswer++;
       }
```
Comment (PoojanPatelJT) [inline src/app/component/quiz/quiz.component.ts:188]: Use Camelcase

File: src/app/component/result/result.component.html [hunk 8]
```diff
@@ -1,66 +1,37 @@
...
       </div>
     </div>
+    </div>
 
     <div class="row" *ngIf="submittedData?.length">
       <div class="col-4 p-3" *ngFor="let data of submittedData">
         <div class="">
           <div class="card shadow">
             <div class="">
               <img class="card-img-top" src="{{ data.image }}" alt="" />
```
Comment (PoojanPatelJT) [inline src/app/component/result/result.component.html]: please check and remove extra div

Comment (PoojanPatelJT) [inline src/app/component/result/result.component.html]: Add proper title inside alt

File: src/app/component/result/result.component.scss [hunk 9]
```diff
@@ -1,60 +1,62 @@
...
+.card-text {
+  font-size: 12px;
+  margin-top: 10px;
+  text-align: left;
+  padding-left: 10px;
 }
-.card-body{
-    height:auto;
+.card-body {
+  height: auto;
```
Comment (PoojanPatelJT) [inline src/app/component/result/result.component.scss]: please use h-auto class

File: src/app/component/startquiz/startquiz.component.ts [hunk 10]
```diff
@@ -1,37 +1,34 @@
...
-import { QuestionService } from 'src/app/service/question.service';
+
 import quizData from '../../../assets/json/data.json';
 
 @Component({
   selector: 'app-carddesign',
   templateUrl: './startquiz.component.html',
-  styleUrls: ['./startquiz.component.scss']
+  styleUrls: ['./startquiz.component.scss'],
 })
```
Comment (PoojanPatelJT) [inline src/app/component/startquiz/startquiz.component.ts:10]: Remove extra comma

File: src/assets/json/data.json [hunk 11]
```diff
@@ -1,1107 +1,1071 @@
...
-          "id": "Q-10.",
           "question": "The statement a===b refers to _________",
           "options": [
             {
-              "value":"a",
-              "lable":"Both a and b are equal in value, type and reference address",
-              "type":"radio"
+              "id": 1,
+              "label": "Both a and b are equal in id, type and reference address",
+              "type": "radio"
```
Comment (PoojanPatelJT) [inline src/assets/json/data.json:632]: please replace id with value

**Review by PoojanPatelJT — CHANGES_REQUESTED** [review]

File: src/assets/json/data.json (hunk 11 above)
Comment (PrabhatJTDeveloper) [reply src/assets/json/data.json:632]: yes i replaced

File: src/app/component/startquiz/startquiz.component.ts (hunk 10 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/startquiz/startquiz.component.ts:10]: yes it is removed

File: src/app/component/result/result.component.scss (hunk 9 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/result/result.component.scss]: yes i sed h-auto case

File: src/app/component/result/result.component.html (hunk 8 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/result/result.component.html]: Yes i removed

Comment (PrabhatJTDeveloper) [reply src/app/component/result/result.component.html]: yes i inserted

File: src/app/component/dashboard/dashboard.component.html [hunk 12]
```diff
@@ -1,32 +1,37 @@
...
-                  class="btn btn-primary"
-                  (click)="quizname(data.quizId)"
-                >
-                  Start Quiz
-                </button>
+          <div class="row">
+            <ng-container *ngFor="let data of quizs; let i = index">
+              <div class="col-lg-3 col-md-6 col-sm-12 col-xs-12">
+                <div class="card shadow">
+                  <img class="card-img-top" src="{{ data.image }}" alt="" />
```
Comment (charmijtdev) [inline src/app/component/dashboard/dashboard.component.html]: please provide alt here too

File: src/app/component/dashboard/dashboard.component.ts [hunk 13]
```diff
@@ -1,36 +1,27 @@
...
+import { Component, OnDestroy, OnInit } from '@angular/core';
 import { Params, Router } from '@angular/router';
 import quizData from '../../../assets/json/data.json';
-import { QuestionService } from '../../service/question.service';
 
 @Component({
   selector: 'app-dashboard',
   templateUrl: './dashboard.component.html',
-  styleUrls: ['./dashboard.component.scss']
+  styleUrls: ['./dashboard.component.scss'],
```
Comment (charmijtdev) [inline src/app/component/dashboard/dashboard.component.ts]: please remove extra comma here too

File: src/app/component/quiz/quiz.component.ts (hunk 7 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/quiz/quiz.component.ts:188]: yes used

File: src/app/component/quiz/quiz.component.ts (hunk 6 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/quiz/quiz.component.ts]: yes used

File: src/app/component/quiz/quiz.component.html [hunk 14]
```diff
@@ -1,53 +1,37 @@
...
-
-        <div formArrayName="form">
-          <ngb-carousel
-            #carousel
-            *ngIf="name"
-            [showNavigationArrows]="false"
-            [showNavigationIndicators]="false"
-            [interval]="0"
-            [keyboard]="false"
+<div class="carousel slide" data-bs-touch="false" data-bs-interval="false">
```
Comment (charmijtdev) [inline src/app/component/quiz/quiz.component.html]: what is use for data-bs-touch and data-bs-interval

Comment (charmijtdev) [reply src/app/component/quiz/quiz.component.html]: also please remove carousal slide too as now we are using ng boostrtap crousla

File: src/app/component/quiz/quiz.component.ts (hunk 5 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/quiz/quiz.component.ts]: yes i used navigateByUrl

File: src/app/component/quiz/quiz.component.html [hunk 15]
```diff
@@ -1,53 +1,37 @@
...
+  <div class="container carousel-inner">
+    <form [formGroup]="myForm">
```
Comment (charmijtdev) [inline src/app/component/quiz/quiz.component.html]: can you pleaase change name for form myForm to quizForm

File: src/app/component/quiz/quiz.component.ts (hunk 4 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/quiz/quiz.component.ts]: removed it because of no use

File: src/app/component/quiz/quiz.component.html (hunk 1 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/quiz/quiz.component.html]: it is already changed in another branch

File: src/app/component/quiz/quiz.component.ts [hunk 16]
```diff
@@ -156,75 +144,67 @@ export class Quizcomponent implements OnInit, OnDestroy {
...
-      if (questionindex === this.question.length) {
-        this.isQuizCompleted = true;
-      }
-      let selectedAnswer =
-        this.question[questionindex].answer?.[0] == correctOptions;
-      if (selectedAnswer && correctOptions) {
-        this.points = this.points += 1;
+  answer(questionIndex: number, correctOptions: string) {
+    if (!this.FormArray.at(questionIndex).get('timer')?.disabled) {
+      if (this.question[questionIndex].answer.id && correctOptions) {
```
Comment (charmijtdev) [inline src/app/component/quiz/quiz.component.ts]: why condition is like this.question[questionIndex].answer.id && correctOption 
 It should be suppose to this.question[questionindex].answer.id == correctOptions 
 And please change correctOptions to correctOption

Comment (charmijtdev) [reply src/app/component/quiz/quiz.component.ts]: Better if we change name to selectedOption

File: src/app/component/startquiz/startquiz.component.ts [hunk 17]
```diff
@@ -1,37 +1,34 @@
...
-  selectedQuizType: any;
-  constructor(
-    private route:Router,
-    private activeRoute: ActivatedRoute,
-    public questionService: QuestionService
-  ){}
-
+  quizData = { ...quizData };
+  instruction: any;
+  selectedQuizType!: string;
```
Comment (charmijtdev) [inline src/app/component/startquiz/startquiz.component.ts:14]: Please add space between here

File: src/app/component/startquiz/startquiz.component.scss [hunk 18]
```diff
@@ -1,39 +1,29 @@
-*{
-    font-family: Roboto;
+* {
```
Comment (charmijtdev) [inline src/app/component/startquiz/startquiz.component.scss]: could you please remove this

**Review by charmijtdev — CHANGES_REQUESTED** [review]

File: src/app/component/quiz/quiz.component.html (hunk 2 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/quiz/quiz.component.html]: yes we can use i for indexing, i changed it with i

File: src/app/component/quiz/quiz.component.ts (hunk 3 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/quiz/quiz.component.ts]: yes i used it directly

File: src/app/component/dashboard/dashboard.component.html (hunk 12 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/dashboard/dashboard.component.html]: yes i provided alt there too.

File: src/app/component/dashboard/dashboard.component.ts (hunk 13 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/dashboard/dashboard.component.ts]: removed it

File: src/app/component/quiz/quiz.component.html (hunk 14 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/quiz/quiz.component.html]: Yes  i removed it

File: src/app/component/quiz/quiz.component.html (hunk 15 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/quiz/quiz.component.html]: yes i changed it

File: src/app/component/quiz/quiz.component.ts (hunk 16 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/quiz/quiz.component.ts]: yes i have changed the name and also wrote the proper condition

File: src/app/component/startquiz/startquiz.component.scss (hunk 18 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/startquiz/startquiz.component.scss]: yes i removed it

File: src/app/component/startquiz/startquiz.component.ts (hunk 17 above)
Comment (PrabhatJTDeveloper) [reply src/app/component/startquiz/startquiz.component.ts:14]: yes i added space between constructor and variables

**Review by PoojanPatelJT — APPROVED** [review]

File: src/app/component/dashboard/dashboard.component.ts (hunk 13 above)
Comment (charmijtdev) [reply src/app/component/dashboard/dashboard.component.ts]: i cant see changes as i can still see comma

File: src/app/component/quiz/quiz.component.html [hunk 19]
```diff
@@ -1,144 +1,108 @@
...
+                  <ng-container *ngFor="let card of question[i].options">
+                    <div class="form-check">
+                      <label
+                        class="form-check-label"
+                        *ngIf="card.type == 'checkbox'"
+                        ><input
+                          type="checkbox"
+                          class="form-check-input"
+                          formControlName="radioValue"
+                          value="{{ card.id }}"
```
Comment (charmijtdev) [inline src/app/component/quiz/quiz.component.html]: please add null check here

File: src/app/component/quiz/quiz.component.html [hunk 20]
```diff
@@ -1,144 +1,108 @@
...
-                        data-bs-slide="per"
-                        style="
-                          position: absolute;
-                          left: 40%;
+                    </div>
+                  </div>
 
-                          width: 20%;
-                        "
+                  <ng-container *ngFor="let card of question[i].options">
```
Comment (charmijtdev) [inline src/app/component/quiz/quiz.component.html]: please add null check here

File: src/app/component/quiz/quiz.component.ts [hunk 21]
```diff
@@ -20,73 +22,42 @@ import { DialogService } from 'src/app/dialog-service/dialog.service';
...
-  findIndex: any;
-  filter: any;
-  filterquestionsindex: any;
-  checkIfselctedanswer: any;
-  userName: any;
-  timer: any;
+  correctAnswer: number = 0;
+  inCorrectAnswer: number = 0;
+  timer: number | undefined;
   selectedQuizType: any;
```
Comment (charmijtdev) [inline src/app/component/quiz/quiz.component.ts]: type should be string not any

File: src/app/component/quiz/quiz.component.ts [hunk 22]
```diff
@@ -20,73 +22,42 @@ import { DialogService } from 'src/app/dialog-service/dialog.service';
...
-  radioValue: any;
+  positivePoints: any;
```
Comment (charmijtdev) [inline src/app/component/quiz/quiz.component.ts]: point should be in number not in any

File: src/app/component/quiz/quiz.component.html [hunk 23]
```diff
@@ -1,144 +1,108 @@
...
-                      </div>
-                      <div class="ram">
-                        <i
-                          style="font-size: 20px; width: 80px"
-                          class="fa fa-clock"
-                          >&#xf017;
-                          {{ FormArray.controls.at(i)?.value?.timer }} sec</i
-                        >
-                      </div>
+<div class="container carousel-inner">
```
Comment (charmijtdev) [inline src/app/component/quiz/quiz.component.html]: please remove this class carousel-inner

File: src/app/component/quiz/quiz.component.html [hunk 24]
```diff
@@ -1,144 +1,108 @@
...
+                      type="button"
+                      (click)="skipQuestion(i)"
+                      class="btn btn-dark skip-button"
+                      aria-hidden="true"
+                      data-bs-slide="per"
+                    >
+                      Skip
+                    </button>
+                    <button
+                      [disabled]="i == 0"
```
Comment (charmijtdev) [inline src/app/component/quiz/quiz.component.html:87]: Please match with length instead 0

File: src/app/component/quiz/quiz.component.ts [hunk 25]
```diff
@@ -116,115 +93,107 @@ export class Quizcomponent implements OnInit, OnDestroy {
...
+  submitQuiz() {
     let data: any = localStorage.getItem('result')
       ? localStorage.getItem('result')
       : [];
     let stringifyData = data.length == 0 ? data : JSON.parse(data);
     let currentData = {
       points: this.points,
-      correctanswer: this.correctanswer,
-      inCorrectAnswer: this.inCorrectanswer,
+      correctanswer: this.correctAnswer,
```
Comment (charmijtdev) [inline src/app/component/quiz/quiz.component.ts]: please change this to correctAnswer

File: src/app/component/quiz/quiz.component.ts [hunk 26]
```diff
@@ -116,115 +93,107 @@ export class Quizcomponent implements OnInit, OnDestroy {
...
       user: this.questionService.getUser(),
     };
     stringifyData.push(currentData);
 
     localStorage.setItem('result', JSON.stringify(stringifyData));
     this.questionService.points = this.points;
-    this.questionService.correctanswer = this.correctanswer;
-    this.questionService.inCorrectAnswer = this.inCorrectanswer;
-    this.router.navigate(['/result']);
+    this.questionService.correctanswer = this.correctAnswer;
```
Comment (charmijtdev) [inline src/app/component/quiz/quiz.component.ts]: no need to add points and all to question service as no longer in use

File: src/app/component/result/result.component.html [hunk 27]
```diff
@@ -1,80 +1,47 @@
...
-                  Go Again To Quiz
-                </button>
-              </div>
-            </div>
+            <span>
+              <h5 class="card-text">Total Points : {{ data.points }}</h5>
+            </span>
+            <span>
+              <h5 class="card-text">
+                Correct Answers : {{ data.correctanswer }}
```
Comment (charmijtdev) [inline src/app/component/result/result.component.html]: change here too

File: src/app/component/result/result.component.html [hunk 28]
```diff
@@ -1,80 +1,47 @@
...
+              <h5 class="card-text">
+                Wrong Answers : {{ data.inCorrectAnswer }}
+              </h5>
+            </span>
+          </p>
+          <div class="text-center">
+            <button
+              type="button"
+              class="btn btn-success"
+              routerLink="/dashboard"
```
Comment (charmijtdev) [inline src/app/component/result/result.component.html]: here go to quiz should be to specific quiz ?

File: src/app/component/startquiz/startquiz.component.html [hunk 29]
```diff
@@ -1,14 +1,19 @@
...
-    <div class="text-center">
-      <button class="btn btn-primary" name="button" (click)="quiz()">
-        Start Quiz
-      </button>
+<div class="container">
+  <div class="d-flex min-vh-100 align-items-center justify-content-around">
+    <div class="card shadow card-layout">
+      <div class="card-body">
+        <div class="container">
+        <img class="card-img-top" src="assets\img\rules.png" alt="" />
```
Comment (charmijtdev) [inline src/app/component/startquiz/startquiz.component.html]: please provide alt here to

File: src/app/component/startquiz/startquiz.component.ts [hunk 30]
```diff
@@ -9,29 +9,27 @@ import quizData from '../../../assets/json/data.json';
...
+  
+  constructor(private route: Router, private activeRoute: ActivatedRoute) {}
 
   ngOnInit(): void {
     this.mapJSONData();
```
Comment (charmijtdev) [inline src/app/component/startquiz/startquiz.component.ts]: please use camelcase type

**Review by charmijtdev — CHANGES_REQUESTED** [review]

**Review by charmijtdev — APPROVED** [review]

**Pull request merged**